File: src/lib/tauri/playlist.ts
import { invoke } from "@tauri-apps/api/core";
import {
  expandTrackResponse,
  type ListTracksResponse,
  type Track,
} from "./track";

/**
 * Playlist row returned by the Rust backend (mirrors
 * `commands::playlist::Playlist`). `track_count` and `total_duration_ms`
 * are computed by the SELECT, not cached columns.
 */
export interface Playlist {
  id: number;
  name: string;
  description: string | null;
  color_id: string;
  icon_id: string;
  is_smart: number;
  /** Blake3 hash of the cover image. Always paired with `cover_path`
   * when the file is materialized; consumers should prefer `cover_path`
   * for rendering and only fall back to `cover_hash` for cache busting. */
  cover_hash: string | null;
  /** Absolute on-disk path resolved by the backend. Pass through
   * `convertFileSrc` to render. `null` means "no custom cover —
   * draw the icon + color gradient instead". */
  cover_path: string | null;
  /** `1` when the cover is auto-managed (regenerates from the first
   * 4 album artworks after every mutation). `0` when the user
   * uploaded their own image; mutations leave it alone until they
   * click "Remove photo" to switch back to auto. */
  cover_is_auto: number;
  position: number;
  created_at: number;
  updated_at: number;
  track_count: number;
  total_duration_ms: number;
  /** Raw JSON payload from `playlist.smart_rules`. `null` for user
   * playlists. For smart playlists the frontend parses the `kind`
   * discriminant to distinguish Daily Mix slots, On Repeat, and
   * custom rule sets — see {@link smartPlaylistKind}. */
  smart_rules: string | null;
}

/** Family discriminator parsed from {@link Playlist.smart_rules}. */
export type SmartPlaylistKind =
  | { kind: "daily_mix"; slot: number }
  | { kind: "on_repeat" }
  | { kind: "custom" }
  | null;

/**
 * Parse a playlist's `smart_rules` JSON into its family discriminant.
 * Returns `null` for user playlists or smart playlists with an
 * unrecognised payload — callers should treat that as "fall back to
 * the generic smart-playlist styling".
 */
export function smartPlaylistKind(p: Playlist): SmartPlaylistKind {
  if (p.is_smart !== 1 || !p.smart_rules) return null;
  try {
    const parsed = JSON.parse(p.smart_rules) as { kind?: string; slot?: number };
    if (parsed.kind === "daily_mix" && typeof parsed.slot === "number") {
      return { kind: "daily_mix", slot: parsed.slot };
    }
    if (parsed.kind === "on_repeat") return { kind: "on_repeat" };
    if (parsed.kind === "custom") return { kind: "custom" };
    return null;
  } catch {
    return null;
  }
}

export interface CreatePlaylistInput {
  name: string;
  description?: string | null;
  color_id?: string;
  icon_id?: string;
}

/** Partial update payload — any omitted field is preserved. */
export interface UpdatePlaylistInput {
  name?: string;
  description?: string | null;
  color_id?: string;
  icon_id?: string;
}

export function listPlaylists(): Promise<Playlist[]> {
  return invoke<Playlist[]>("list_playlists");
}

export function getPlaylist(playlistId: number): Promise<Playlist> {
  return invoke<Playlist>("get_playlist", { playlistId });
}

export function createPlaylist(input: CreatePlaylistInput): Promise<Playlist> {
  return invoke<Playlist>("create_playlist", { input });
}

export function updatePlaylist(
  playlistId: number,
  input: UpdatePlaylistInput,
): Promise<void> {
  return invoke<void>("update_playlist", { playlistId, input });
}

export function deletePlaylist(playlistId: number): Promise<void> {
  return invoke<void>("delete_playlist", { playlistId });
}

export async function listPlaylistTracks(playlistId: number): Promise<Track[]> {
  const resp = await invoke<ListTracksResponse>("list_playlist_tracks", {
    playlistId,
  });
  return expandTrackResponse(resp);
}

export function addTrackToPlaylist(
  playlistId: number,
  trackId: number,
): Promise<void> {
  return invoke<void>("add_track_to_playlist", { playlistId, trackId });
}

export function addTracksToPlaylist(
  playlistId: number,
  trackIds: number[],
): Promise<number> {
  return invoke<number>("add_tracks_to_playlist", { playlistId, trackIds });
}

export function removeTrackFromPlaylist(
  playlistId: number,
  trackId: number,
): Promise<void> {
  return invoke<void>("remove_track_from_playlist", { playlistId, trackId });
}

/**
 * Return the IDs of every user playlist that currently contains the track.
 * Smart playlists are excluded — their membership is rule-driven and not
 * a toggle target. Used by the `+` popover to mark existing memberships
 * with a checkmark and switch the click handler from add to remove.
 */
export function listPlaylistsContainingTrack(
  trackId: number,
): Promise<number[]> {
  return invoke<number[]>("list_playlists_containing_track", { trackId });
}

export function reorderPlaylistTrack(
  playlistId: number,
  trackId: number,
  newPosition: number,
): Promise<void> {
  return invoke<void>("reorder_playlist_track", {
    playlistId,
    trackId,
    newPosition,
  });
}

/**
 * Add every track belonging to a source (folder, album, artist) to a
 * playlist. The SELECT + INSERT runs server-side in one transaction so
 * no track IDs travel through the IPC bridge.
 */
export function addSourceToPlaylist(
  playlistId: number,
  sourceType: "folder" | "album" | "artist",
  sourceId: number,
): Promise<number> {
  return invoke<number>("add_source_to_playlist", {
    playlistId,
    sourceType,
    sourceId,
  });
}

export interface ImportPlaylistResult {
  playlist_id: number;
  imported: number;
  missing: number;
  /** Up to 20 unmatched paths from the imported file. */
  missing_paths: string[];
}

/** Write the playlist out as a UTF-8 .m3u8 file at `destPath`. */
export function exportPlaylistM3u(
  playlistId: number,
  destPath: string,
): Promise<void> {
  return invoke<void>("export_playlist_m3u", { playlistId, destPath });
}

/**
 * Parse an .m3u/.m3u8 file, match its entries against the active
 * library, and create a new playlist holding the resolved tracks.
 */
export function importPlaylistM3u(
  sourcePath: string,
): Promise<ImportPlaylistResult> {
  return invoke<ImportPlaylistResult>("import_playlist_m3u", { sourcePath });
}

/**
 * Upload a user-supplied image as the playlist cover. Validates the file
 * by magic bytes (jpg/png/webp), normalises through the same compositor
 * used for auto-covers (re-encodes to a 640×640 JPEG), and flips
 * `cover_is_auto` to 0 so subsequent mutations stop overwriting it.
 */
export function setPlaylistCoverFromFile(
  playlistId: number,
  filePath: string,
): Promise<void> {
  return invoke<void>("set_playlist_cover_from_file", { playlistId, filePath });
}

/**
 * Force a regen of the auto-cover (Spotify-style 2×2 grid of the first
 * 4 album artworks). Normally the auto pipeline runs implicitly after
 * every mutation; this command is the "refresh now" escape hatch.
 */
export function regeneratePlaylistAutoCover(playlistId: number): Promise<void> {
  return invoke<void>("regenerate_playlist_auto_cover", { playlistId });
}

/**
 * Drop the manual cover and switch back to auto mode. Immediately
 * re-runs the auto-cover so the visual feedback is instant.
 */
export function clearPlaylistCover(playlistId: number): Promise<void> {
  return invoke<void>("clear_playlist_cover", { playlistId });
}

File: src-tauri/crates/core/src/domain/playlist.rs
//! Playlist DTOs. Covers user-curated playlists and the smart-playlist
//! family (Daily Mix slots, On Repeat, custom rule trees) — the
//! distinction is carried by `is_smart` + the JSON blob in
//! `smart_rules`, which the frontend parses to dispatch on the
//! discriminant.

use serde::{Deserialize, Serialize};

/// Playlist row returned to the frontend, with the track count + total
/// duration denormalised so the sidebar can show "Playlist · N titres"
/// without issuing a follow-up query per playlist.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(any(feature = "sqlite", feature = "postgres"), derive(sqlx::FromRow))]
pub struct Playlist {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub color_id: String,
    pub icon_id: String,
    pub is_smart: i64,
    /// Blake3 hash of the cover image stored in the shared
    /// `metadata_artwork` cache. `None` for user-curated playlists that
    /// haven't been given a custom cover — the frontend renders the
    /// `icon_id` + `color_id` gradient instead. Smart playlists populate
    /// this with a composite generated from the cluster's top artists.
    pub cover_hash: Option<String>,
    /// Absolute on-disk path resolved from `cover_hash` if the file is
    /// actually present, ready to be passed to `convertFileSrc`. The
    /// frontend prefers this over re-resolving the hash itself so a
    /// stale `cover_hash` (cache wiped) doesn't render a broken image.
    pub cover_path: Option<String>,
    /// `1` when the cover is managed by the auto-regen pipeline
    /// (Spotify-style 2×2 grid of the first 4 album artworks; refreshed
    /// after every mutation that could change the first-4 set). `0` when
    /// the user uploaded their own image — playlist mutations then leave
    /// the cover alone.
    pub cover_is_auto: i64,
    pub position: i64,
    pub created_at: i64,
    pub updated_at: i64,
    pub track_count: i64,
    pub total_duration_ms: i64,
    /// Raw JSON payload from `playlist.smart_rules`. `None` for user
    /// playlists (`is_smart = 0`); for smart playlists the frontend
    /// parses the `kind` discriminant to distinguish Daily Mix slots,
    /// On Repeat, and custom rule sets without an extra round-trip.
    pub smart_rules: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreatePlaylistInput {
    pub name: String,
    pub description: Option<String>,
    pub color_id: Option<String>,
    pub icon_id: Option<String>,
}

/// Partial update payload — any field left as `None` is preserved via
/// SQL `COALESCE`. Same shape as [`crate::domain::library::UpdateLibraryInput`].
#[derive(Debug, Deserialize)]
pub struct UpdatePlaylistInput {
    pub name: Option<String>,
    pub description: Option<String>,
    pub color_id: Option<String>,
    pub icon_id: Option<String>,
}
